modules/graceful/server.go

```diff
@@ -0,0 +1,267 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+// This code is highly inspired by endless go
+
+package graceful
+
+import (
+	"crypto/tls"
+	"net"
+	"os"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"code.gitea.io/gitea/modules/log"
+)
+
+type state uint8
+
+const (
+	stateInit state = iota
+	stateRunning
+	stateShuttingDown
+	stateTerminate
+)
+
+var (
+	// RWMutex for when adding servers or shutting down
+	runningServerReg sync.RWMutex
+	// ensure we only fork once
+	runningServersForked bool
+
+	// DefaultReadTimeOut default read timeout
+	DefaultReadTimeOut time.Duration
+	// DefaultWriteTimeOut default write timeout
+	DefaultWriteTimeOut time.Duration
+	// DefaultMaxHeaderBytes default max header bytes
+	DefaultMaxHeaderBytes int
+
+	// IsChild reports if we are a fork iff LISTEN_FDS is set and our parent PID is not 1
+	IsChild = len(os.Getenv(listenFDs)) > 0 && os.Getppid() > 1
+)
+
+func init() {
+	runningServerReg = sync.RWMutex{}
+
+	DefaultMaxHeaderBytes = 0 // use http.DefaultMaxHeaderBytes - which currently is 1 << 20 (1MB)
+}
+
+// ServeFunction represents a listen.Accept loop
+type ServeFunction = func(net.Listener) error
+
+// Server represents our graceful server
+type Server struct {
+	network         string
+	address         string
+	listener        net.Listener
+	PreSignalHooks  map[os.Signal][]func()
+	PostSignalHooks map[os.Signal][]func()
+	wg              sync.WaitGroup
+	sigChan         chan os.Signal
+	state           state
+	lock            *sync.RWMutex
+	BeforeBegin     func(network, address string)
+	OnShutdown      func()
+}
+
+// NewServer creates a server on network at provided address
+func NewServer(network, address string) *Server {
+	runningServerReg.Lock()
+	defer runningServerReg.Unlock()
+
+	if IsChild {
+		log.Info("Restarting new server: %s:%s on PID: %d", network, address, os.Getpid())
+	} else {
+		log.Info("Starting new server: %s:%s on PID: %d", network, address, os.Getpid())
+	}
+	srv := &Server{
+		wg:              sync.WaitGroup{},
+		sigChan:         make(chan os.Signal),
+		PreSignalHooks:  map[os.Signal][]func(){},
+		PostSignalHooks: map[os.Signal][]func(){},
+		state:           stateInit,
+		lock:            &sync.RWMutex{},
+		network:         network,
+		address:         address,
+	}
+
+	srv.BeforeBegin = func(network, addr string) {
+		log.Debug("Starting server on %s:%s (PID: %d)", network, addr, syscall.Getpid())
+	}
+
+	return srv
+}
+
+// ListenAndServe listens on the provided network address and then calls Serve
+// to handle requests on incoming connections.
+func (srv *Server) ListenAndServe(serve ServeFunction) error {
+	go srv.handleSignals()
+
+	l, err := GetListener(srv.network, srv.address)
```

Comment:
I don't think we can detect which sockets are unused. Sockets are not all opened in one place.

Reply:
I mean those passed on with `LISTEN_FDS`. We should close those that weren't picked up.

Reply:
As far as I'm aware there's no "choke point" where we have finished opening all of the listeners, except maybe just before shutdown which is not really helpful as they're gonna be closed imminently by the process shutting down.

The only time I could imagine it mattering is when you change the listening ports during graceful restart - you'll get the old port being passed and held as well as the new one being opened. If you restart the process again then the old one should then be dropped. 

Clearly that's not ideal. 

I'll have a think about what we could do. 

Maybe add another waitgroup and goroutine that could then close the unused fds? We'd have a fixed number waits (3 at present for http, https and SSH) and once they're done deciding to be configured or not we can then let the goroutine close the rest of the fds. Hmm... That could work - the only issue is that if we add anything else that could listen we'll have to remember to increase the hard coded number as I don't think there's any way we can register these.

Reply:
OK I've added a cleanup routine - it's slightly ugly IMO but it's the only way we can do it.

I've also added a note on `routers/install.go` which could be fixed in a subsequent PR as the primary problem is not a graceful issue per se.

Reply:
Sorry, that's not what I meant. I meant something much simpler.
When a new instance is started from the gracefull cmd, you repeatedly call `GetListener()` in the initialization consuming sockets from `LISTEN_FDS`, which is currently a slice. You can just make a that map instead of a slice and delete the sockets from it as you find a match in `GetListener()`. So, at the end of your initialization, when there's no more `GetListener()` to call, those entries remaining in the map are not needed anymore and can be closed.

Reply:
I don't understand.

As I said above you can't tell what you're going to need until after you've started the services and they all start in different goroutines. There's no synchronous system that parses our config, does setup and then starts stuff - setup and service start occurs at the same point.

Reply:
OK, I see. It's really an edge case for the user to need the listening port for anything else.

Reply:
Sorry I made you work so hard for this. I really thought it was simpler. Anyway, now it's covered. 👍

Reply:
No worries. The changes are an improvement on what I had originally and have got me to think about the problems of install.go a little. This will at least lead to release of unused listeners now whereas previously it couldn't.